### PR TITLE
Post processing defences

### LINKED ISSUE #267: Move post-processing defences to defences and integrate with classifier
Move post-processing defences from module wrappers to defences and integrate them similar to pre-processing defences into the classifiers.

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
Fixes #267

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/classifiers/GPy.py [hunk 1]
```diff
@@ -141,18 +144,24 @@ def predict(self, x, logits=False, **kwargs):
         :return: Array of predictions of shape `(nb_inputs, nb_classes)`.
         :rtype: `np.ndarray`
         """
-        # Apply preprocessing
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]:
```suggestion
        # Apply preprocessing
```

File: tests/test_utils.py [hunk 2]
```diff
@@ -348,6 +349,13 @@ def first_class(true_labels, model_labels, idx=0):
         self.assertEqual(performance_diff(full_model, limited_model, x_test, y_test, perf_function=first_class, idx=1),
                          -1.0 / 3)
 
+    def test_is_probability(self):
+        x = np.array([0.1, 0.3, 0.5999])
+        y = np.array([0.1, 0.3, 0.599])
+
```
Comment (beat-buesser) [inline tests/test_utils.py]: Please add `assertFalse` tests for negative values and for values larger than one

File: art/utils.py [hunk 3]
```diff
@@ -782,3 +782,16 @@ def performance_diff(model1, model2, test_data, test_labels, perf_function='accu
...
+
+def is_probability(vector):
+    """
+    Check if an 1D-array is a probability vector.
+
+    :param vector: An 1D-array.
+    :type vector: `np.ndarray`
+    :return: True if it is a probability vector.
+    :rtype: `bool`
+    """
```
Comment (beat-buesser) [inline art/utils.py]: Please add checks for negative values or values larger than 1.

File: art/defences/postprocess/postprocessor.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+This module implements the abstract base class for defences that post-process classifier output.
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import abc
+
+
+class Postprocessor(abc.ABC):
+    """
+    Abstract base class for defences performing model hardening by postprocessing model output.
```
Comment (beat-buesser) [inline art/defences/postprocess/postprocessor.py]:
```suggestion
    Abstract base class for postprocessing defences.
```

File: art/defences/preprocess/preprocessor.py [hunk 5]
```diff
@@ -21,16 +21,9 @@
...
-if sys.version_info >= (3, 4):
-    ABC = abc.ABC
-else:
-    ABC = abc.ABCMeta(str('ABC'), (), {})
 
-
-class Preprocessor(ABC):
+class Preprocessor(abc.ABC):
     """
     Abstract base class for defences performing model hardening by preprocessing data.
```
Comment (beat-buesser) [inline art/defences/preprocess/preprocessor.py]:
```suggestion
    Abstract base class for preprocessing defences.
```

File: art/classifiers/GPy.py [hunk 6]
```diff
@@ -141,18 +144,24 @@ def predict(self, x, logits=False, **kwargs):
...
-        # output normal prediction, scale up to two values
-        out[:, 0] = self.model.predict(x_preprocessed)[0].reshape(-1)
-        out[:, 1] = 1.0 - out[:, 0]
-        return out
+        else:
+            # output normal prediction, scale up to two values
+            out[:, 0] = self.model.predict(x_preprocessed)[0].reshape(-1)
+            out[:, 1] = 1.0 - out[:, 0]
+
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/GPy.py [hunk 7]
```diff
@@ -163,10 +172,16 @@ def predict_uncertainty(self, x):
         :return: Array of uncertainty predictions of shape `(nb_inputs)`.
         :rtype: `np.ndarray`
         """
-        # Apply preprocessing
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/GPy.py [hunk 8]
```diff
@@ -163,10 +172,16 @@ def predict_uncertainty(self, x):
...
         x_preprocessed, _ = self._apply_preprocessing(x, y=None, fit=False)
 
-        return self.model.predict_noiseless(x_preprocessed)[1]
+        # Perform prediction
+        out = self.model.predict_noiseless(x_preprocessed)[1]
+
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/GPy.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/blackbox.py [hunk 9]
```diff
@@ -77,7 +82,7 @@ def predict(self, x, batch_size=128, **kwargs):
         """
         from art.config import ART_NUMPY_DTYPE
 
-        # Apply defences
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/blackbox.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/blackbox.py [hunk 10]
```diff
@@ -86,6 +91,9 @@ def predict(self, x, batch_size=128, **kwargs):
             begin, end = batch_index * batch_size, min((batch_index + 1) * batch_size, x_preprocessed.shape[0])
             predictions[begin:end] = self._predictions(x_preprocessed[begin:end])
 
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/blackbox.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/catboost.py [hunk 11]
```diff
@@ -88,10 +95,16 @@ def predict(self, x, **kwargs):
         :return: Array of predictions of shape `(nb_inputs, nb_classes)`.
         :rtype: `np.ndarray`
         """
-        # Apply defences
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/catboost.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/catboost.py [hunk 12]
```diff
@@ -88,10 +95,16 @@ def predict(self, x, **kwargs):
...
         x_preprocessed, _ = self._apply_preprocessing(x, y=None, fit=False)
 
-        return self._model.predict_proba(x_preprocessed)
+        # Perform prediction
+        predictions = self._model.predict_proba(x_preprocessed)
+
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/catboost.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/detector_classifier.py [hunk 13]
```diff
@@ -38,23 +38,30 @@ class DetectorClassifier(ClassifierNeuralNetwork, ClassifierGradients, Classifie
...
-        :type defences: `str` or `list(str)`
+        :param preprocessing_defences: Preprocessing defence(s) to be activated with the classifier. Not applicable
+               in this classifier.
+        :type preprocessing_defences: :class:`.Preprocessor` or `list(Preprocessor)` instances
+        :param postprocessing_defences: Postprocessing defence(s) to be activated with the classifier.
+        :type postprocessing_defences: :class:`.Postprocessor` or `list(Postprocessor)` instances
         :param preprocessing: Tuple of the form `(subtractor, divider)` of floats or `np.ndarray` of values to be
                used for data preprocessing. The first value will be subtracted from the input. The input will then
-               be divided by the second one.
+               be divided by the second one. Not applicable in this classifier.
```
Comment (beat-buesser) [inline art/classifiers/detector_classifier.py]: Should we raise an exception if preprocessing is not None?

File: art/classifiers/detector_classifier.py [hunk 14]
```diff
@@ -73,17 +80,17 @@ def predict(self, x, batch_size=128, **kwargs):
...
-        classifier_outputs = self.classifier.predict(x=x_defences, batch_size=batch_size)
-        detector_outputs = self.detector.predict(x=x_defences, batch_size=batch_size)
+        classifier_outputs = self.classifier.predict(x=x, batch_size=batch_size)
+        detector_outputs = self.detector.predict(x=x, batch_size=batch_size)
         detector_outputs = (np.reshape(detector_outputs, [-1]) + 1) * np.max(classifier_outputs, axis=1)
         detector_outputs = np.reshape(detector_outputs, [-1, 1])
         combined_outputs = np.concatenate([classifier_outputs, detector_outputs], axis=1)
 
-        return combined_outputs
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/detector_classifier.py]:
```suggestion
        # Apply postprocessing
```

File: art/defences/postprocess/postprocessor.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+# SOFTWARE.
+"""
+This module implements the abstract base class for defences that post-process classifier output.
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import abc
+
+
+class Postprocessor(abc.ABC):
```
Comment (beat-buesser) [inline art/defences/postprocess/postprocessor.py]: I think we also need to add
```
@abc.abstractmethod
    def estimate_gradient(self, x, grad):
```
for postprocessors.

File: art/classifiers/classifier.py [hunk 16]
```diff
@@ -275,6 +279,29 @@ def _apply_preprocessing_standardisation(self, x):
 
         return res
 
+    def _apply_postprocessing(self, preds, fit):
```
Comment (beat-buesser) [inline art/classifiers/classifier.py]: I think we need to add a call to `postprocessor.estimate_gradient`.

File: art/classifiers/keras.py [hunk 17]
```diff
@@ -304,7 +309,7 @@ def predict(self, x, batch_size=128, **kwargs):
         """
         from art.config import ART_NUMPY_DTYPE
 
-        # Apply defences
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/keras.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/keras.py [hunk 18]
```diff
@@ -313,6 +318,9 @@ def predict(self, x, batch_size=128, **kwargs):
             begin, end = batch_index * batch_size, min((batch_index + 1) * batch_size, x_preprocessed.shape[0])
             predictions[begin:end] = self._predictions([x_preprocessed[begin:end]])[0]
 
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/keras.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/lightgbm.py [hunk 19]
```diff
@@ -87,10 +92,16 @@ def predict(self, x, **kwargs):
         :return: Array of predictions of shape `(nb_inputs, nb_classes)`.
         :rtype: `np.ndarray`
         """
-        # Apply defences
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/lightgbm.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/lightgbm.py [hunk 20]
```diff
@@ -87,10 +92,16 @@ def predict(self, x, **kwargs):
...
         x_preprocessed, _ = self._apply_preprocessing(x, y=None, fit=False)
 
-        return self._model.predict(x_preprocessed)
+        # Perform prediction
+        predictions = self._model.predict(x_preprocessed)
+
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/lightgbm.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/ensemble.py [hunk 21]
```diff
@@ -53,15 +53,20 @@ def __init__(self, classifiers, classifier_weights=None, channel_index=3, clip_v
...
-        :type defences: `str` or `list(str)`
+        :param preprocessing_defences: Preprocessing defence(s) to be activated with the classifier. Not applicable
+               in this classifier.
+        :type preprocessing_defences: :class:`.Preprocessor` or `list(Preprocessor)` instances
+        :param postprocessing_defences: Postprocessing defence(s) to be activated with the classifier.
+        :type postprocessing_defences: :class:`.Postprocessor` or `list(Postprocessor)` instances
         :param preprocessing: Tuple of the form `(subtractor, divider)` of floats or `np.ndarray` of values to be
                used for data preprocessing. The first value will be subtracted from the input. The input will then
-               be divided by the second one.
+               be divided by the second one. Not applicable in this classifier.
```
Comment (beat-buesser) [inline art/classifiers/ensemble.py]: Should we raise and exception if preprocessing is not None?

File: art/classifiers/ensemble.py [hunk 22]
```diff
@@ -124,7 +129,11 @@ def predict(self, x, batch_size=128, raw=False, **kwargs):
 
         # Aggregate predictions only at probabilities level, as logits are not comparable between models
         var_z = np.sum(preds, axis=0)
-        return var_z
+
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/ensemble.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/mxnet.py [hunk 23]
```diff
@@ -114,7 +119,7 @@ def fit(self, x, y, batch_size=128, nb_epochs=20, **kwargs):
 
         train_mode = self._learning_phase if hasattr(self, '_learning_phase') else True
 
-        # Apply preprocessing
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/mxnet.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/mxnet.py [hunk 24]
```diff
@@ -133,8 +138,15 @@ def fit(self, x, y, batch_size=128, nb_epochs=20, **kwargs):
                 y_batch = mx.nd.array(y_preprocessed[ind[m * batch_size:(m + 1) * batch_size]]).as_in_context(self._ctx)
 
                 with mx.autograd.record(train_mode=train_mode):
+                    # Perform prediction
                     preds = self._model(x_batch)
+
+                    # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/mxnet.py]:
```suggestion
                    # Apply postprocessing
```

File: art/classifiers/mxnet.py [hunk 25]
```diff
@@ -169,8 +181,15 @@ def fit_generator(self, generator, nb_epochs=20, **kwargs):
                     y_batch = mx.nd.array(y_batch).as_in_context(self._ctx)
 
                     with mx.autograd.record(train_mode=train_mode):
+                        # Perform prediction
                         preds = self._model(x_batch)
+
+                        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/mxnet.py]:
```suggestion
                        # Apply postprocessing
```

File: art/classifiers/mxnet.py [hunk 26]
```diff
@@ -194,7 +213,7 @@ def predict(self, x, batch_size=128, **kwargs):
 
         train_mode = self._learning_phase if hasattr(self, '_learning_phase') else False
 
-        # Apply preprocessing
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/mxnet.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/mxnet.py [hunk 27]
```diff
@@ -212,7 +231,10 @@ def predict(self, x, batch_size=128, **kwargs):
 
             results[begin:end] = preds.asnumpy()
 
-        return results
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/mxnet.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/pytorch.py [hunk 28]
```diff
@@ -102,7 +106,7 @@ def predict(self, x, batch_size=128, **kwargs):
         """
         import torch
 
-        # Apply preprocessing
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/pytorch.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/pytorch.py [hunk 29]
```diff
@@ -116,7 +120,10 @@ def predict(self, x, batch_size=128, **kwargs):
             output = model_outputs[-1]
             results[begin:end] = output.detach().cpu().numpy()
 
-        return results
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/pytorch.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/pytorch.py [hunk 30]
```diff
@@ -138,7 +145,7 @@ def fit(self, x, y, batch_size=128, nb_epochs=10, **kwargs):
         """
         import torch
 
-        # Apply preprocessing
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/pytorch.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/scikitlearn.py [hunk 31]
```diff
@@ -130,7 +140,10 @@ def predict(self, x, **kwargs):
         else:
             raise ValueError('The provided model does not have methods `predict_proba` or `predict`.')
 
-        return y_pred
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/scikitlearn.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/tensorflow.py [hunk 32]
```diff
@@ -111,7 +116,7 @@ def predict(self, x, batch_size=128, **kwargs):
         :return: Array of predictions of shape `(num_inputs, nb_classes)`.
         :rtype: `np.ndarray`
         """
-        # Apply preprocessing
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/tensorflow.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/tensorflow.py [hunk 33]
```diff
@@ -128,7 +133,10 @@ def predict(self, x, batch_size=128, **kwargs):
             # Run prediction
             results[begin:end] = self._sess.run(self._output, feed_dict=feed_dict)
 
-        return results
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/tensorflow.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/tensorflow.py [hunk 34]
```diff
@@ -670,7 +680,10 @@ def predict(self, x, batch_size=128, **kwargs):
             # Run prediction
             results[begin:end] = self._model(x_preprocessed[begin:end])
 
-        return results
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/tensorflow.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/xgboost.py [hunk 35]
```diff
@@ -96,7 +99,7 @@ def predict(self, x, **kwargs):
         from xgboost import Booster, XGBClassifier
         from art.utils import to_categorical
 
-        # Apply defences
+        # Apply preprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/xgboost.py]:
```suggestion
        # Apply preprocessing
```

File: art/classifiers/xgboost.py [hunk 36]
```diff
@@ -106,12 +109,13 @@ def predict(self, x, **kwargs):
             y_prediction = np.asarray([line for line in predictions])
             if len(y_prediction.shape) == 1:
                 y_prediction = to_categorical(labels=y_prediction, nb_classes=self.nb_classes())
-            return y_prediction
+        else:
+            y_prediction = self._model.predict_proba(x_preprocessed)
 
-        if isinstance(self._model, XGBClassifier):
-            return self._model.predict_proba(x_preprocessed)
+        # Apply postprocessing defences
```
Comment (beat-buesser) [inline art/classifiers/xgboost.py]:
```suggestion
        # Apply postprocessing
```

File: art/classifiers/xgboost.py [hunk 37]
```diff
@@ -106,12 +109,13 @@ def predict(self, x, **kwargs):
             y_prediction = np.asarray([line for line in predictions])
             if len(y_prediction.shape) == 1:
                 y_prediction = to_categorical(labels=y_prediction, nb_classes=self.nb_classes())
-            return y_prediction
+        else:
```
Comment (beat-buesser) [inline art/classifiers/xgboost.py]:
```suggestion
        elif isinstance(self._model, XGBClassifier)::
```

File: tests/test_utils.py (hunk 2 above)
Comment (minhitbk) [reply tests/test_utils.py]: Solved.

File: art/utils.py (hunk 3 above)
Comment (minhitbk) [reply art/utils.py]: Solved.

File: art/classifiers/detector_classifier.py (hunk 13 above)
Comment (minhitbk) [reply art/classifiers/detector_classifier.py]: Solved.

File: art/classifiers/ensemble.py (hunk 21 above)
Comment (minhitbk) [reply art/classifiers/ensemble.py]: Solved.

File: art/defences/postprocess/postprocessor.py (hunk 15 above)
Comment (minhitbk) [reply art/defences/postprocess/postprocessor.py]: as per discussion, we dont need this.

File: art/classifiers/classifier.py (hunk 16 above)
Comment (minhitbk) [reply art/classifiers/classifier.py]: as per discussion, we dont need this.

**Comment by minhitbk** [comment]
I have updated docstrings as well.
